envoy/upstream/outlier_detection.h

```diff
@@ -47,6 +49,104 @@ enum class Result {
   ExtOriginRequestSuccess // Request was completed successfully.
 };
 
+// Types of errors reported to outlier detectors.
+// Each type may have a different syntax and content.
+// TODO (cpakulski) - maybe use enums from protobufs.
+enum class ExtResultType {
+  TEST, // Used in unit tests.
+  HTTP_CODE,
+  LOCAL_ORIGIN,
+};
+
+/*
+ * Class carries result of a transaction with upstream entity
+ * or generated internally by Envoy.
+ * Different categories of results will be derived from that base class.
+ * Those categories of results are fed only into Outlier Detection extensions.
+ */
+class ExtResult {
+public:
+  ExtResult() = delete;
+  ExtResult(ExtResultType type) : type_(type) {}
+  virtual ExtResultType type() const { return type_; };
+  virtual ~ExtResult() = default;
+
+private:
+  const ExtResultType type_;
+};
+
+// Types derived from TypedExtResult are used to report the result of transaction to
+// an outlier detection monitor.
+template <ExtResultType E> class TypedExtResult : public ExtResult {
+protected:
+  TypedExtResult() : ExtResult(E) {}
+};
+
+class HttpCode : public TypedExtResult<ExtResultType::HTTP_CODE> {
+public:
+  HttpCode(uint32_t code) : code_(code) {}
+  HttpCode() = delete;
+  virtual ~HttpCode() {}
+  uint32_t code() const { return code_; }
+
+private:
+  uint32_t code_;
+};
+
+// LocalOriginEvent is used to report errors like resets, timeouts but also
+// successful connection attempts.
+class LocalOriginEvent : public TypedExtResult<ExtResultType::LOCAL_ORIGIN> {
+public:
+  LocalOriginEvent(Result result) : result_(result) {}
+  LocalOriginEvent() = delete;
+  Result result() const { return result_; }
+
+private:
+  Result result_;
+};
```

Comment:
Removed template and used simple inheritance. Thanks!

Reply:
WDYT about to use `absl::variant<HttpCode, LocalOriginEvent>` to avoid the virtual base class and possible `dynamic_cast`?

Reply:
Converted to absl::variant. For a while I though it would not be possible, but actually was easier than I though. Code is even cleaner as I removed enums used to distinguish each type of ExtResult. Great suggestion. Thanks!